File: mergesort.go
package main

import "fmt"

func mergesort(arr []int) []int {
	arrLen := len(arr)
	if arrLen <= 1 {
		return arr
	}

	mid := arrLen / 2
	left := []int(arr[0:mid])
	right := []int(arr[mid:])

	left = mergesort(left)
	right = mergesort(right)

	arr = make([]int, len(left)+len(right))
	var j, k int
	for i := 0; i < len(arr); i++ {
		switch {
		case j >= len(left):
			arr[i] = right[k]
			k++
		case k >= len(right):
			arr[i] = left[j]
			j++
		case left[j] > right[k]:
			arr[i] = right[k]
			k++
		default:
			arr[i] = left[j]
			j++
		}
	}
	return arr
}

func main() {
	a := []int{5, 9, 1, 3, 4, 6, 6, 3, 2}
	fmt.Printf("%v\n", mergesort(a))
}

File: Mergesort.java
import java.util.Arrays;

public class Mergesort {
    
    public static int[] sort(int arr[]) {
        int n = arr.length;
        if (n == 1) {
            return arr;
        }

        int[] left = Arrays.copyOfRange(arr, 0, n / 2);
        int[] right = Arrays.copyOfRange(arr, n / 2, n);
        left = sort(left);
        right = sort(right);
        return merge(left, right);
    }

    public static int[] merge(int[] left, int[] right) {
        int[] arr = new int[left.length + right.length];
        int l = 0, r = 0, i = 0;
        
        while (l < left.length && r < right.length) {
            if (left[l] <= right[r]) {
                arr[i++] = left[l++];
            } else {
                arr[i++] = right[r++];
            }
        }

        while (l < left.length) {
            arr[i++] = left[l++];
        }
        
        while (r < right.length) {
            arr[i++] = right[r++];
        }

        return arr;
    }

    public static void main(String[] args) {
        int[] arr = {5,9,1,3,4,6,6,3,2};
        System.out.println(Arrays.toString(sort(arr)));
    }
}